MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReaderUtils.java

```diff
@@ -0,0 +1,131 @@
+     * extractor.
+     * @param extractor An extractor connected to the data source from which we want to read.
+     * @param decoder A decoder used to decode samples read from the extractor.
+     * @param inputBuffer The input buffer to read samples to and queue to the MediaCodec.
+     * @param bufferIndex The index of the input buffer.
+     */
+    public static void readFromVideoTrack(MediaExtractor extractor, MediaCodec decoder,
+                                          ByteBuffer inputBuffer, int bufferIndex) {
+        int sampleSize = extractor.readSampleData(inputBuffer, 0);
+        if (sampleSize < 0) {
+            decoder.queueInputBuffer(
+                    bufferIndex,
+                    /* offset = */ 0,
+                    /* size = */ 0,
+                    /* presentationTimeUs = */ 0,
+                    MediaCodec.BUFFER_FLAG_END_OF_STREAM
+            );
+        } else {
+            decoder.queueInputBuffer(
+                    bufferIndex,
+                    /* offset = */ 0,
+                    sampleSize,
+                    extractor.getSampleTime(),
+                    /* flags = */ 0
+            );
+        }
+    }
+
+    /**
+     * Retrieve the information of the next available output buffer.
+     * @param availableOutputBuffers A blocking queue of Bundle objects containing information about
+     * free output buffers.
+     * @return the bundle containing the information of the next available output buffer, or null if
+     * the poll call results in a timeout.
+     */
+    public static Bundle getOutputBuffer(BlockingQueue<Bundle> availableOutputBuffers) {
+        Bundle bufferData = null;
+        try {
+            bufferData = availableOutputBuffers.poll(TIMEOUT_US, TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            Log.e(TAG, "No output buffer available", e);
+        }
+        return bufferData;
+    }
+
+    /**
+     * Retrieve the stabilization homographies from the motion track. If the frame has already been
+     * stabilized, then return an empty list.
+     * @param extractor The extractor connected to the track containing the stabilization data.
+     * @param inputBuffer A free input buffer for the extractor to read data into.
+     * @return a List of HomographyMatrix objects.
+     */
+    public static List<HomographyMatrix> getHomographies(MediaExtractor extractor,
+                                                         ByteBuffer inputBuffer) {
+        List<HomographyMatrix> homographyList = new ArrayList<>();
+        int sampleSize = extractor.readSampleData(inputBuffer, 0);
+        if (sampleSize >= 0) {
+            // Deserialize data
+            Stabilization.Data stabilizationData = null;
+            try {
+                stabilizationData = Stabilization.Data.parseFrom(inputBuffer);
+                if (stabilizationData == null) {
+                    throw new RuntimeException("Protocol buffer is null");
```

Comment:
Maybe we would prefer to not crash the entire app in that case? Is there any way that we could "recover" from it? (eg. ignore the stabilization data in that case).

Reply:
(especially since you actually do that for the InvalidProtocolBufferException error a few lines below)

Reply:
Made a boolean to catch if the protocol buffer is null, and if it's null then I just returned a homography list of identity matrices (effectively turning stabilization off).